pennylane/interfaces/batch/jax.py

```diff
@@ -1,299 +1,332 @@
+            # dtype=float32), DeviceArray(0., dtype=float32)]].
+            need_unwrapping = any(r.ndim != 0 for r in res)
+            if need_unwrapping:
+                unwrapped_res = []
+                for r in res:
+                    if r.ndim != 0:
+                        r = [jnp.array(p) for p in r]
+                    unwrapped_res.append(r)
+
+                res = unwrapped_res
+
+            return (tuple(res),)
+
+        # Gradient function is a device method.
+        with qml.tape.Unwrap(*tapes):
+            jacs = gradient_fn(tapes, **gradient_kwargs)
+
+        vjps = [qml.gradients.compute_vjp(d, jac) for d, jac in zip(g, jacs)]
+        res = [[jnp.array(p) for p in v] for v in vjps]
+        return (tuple(res),)
+
+    wrapped_exec.defvjp(wrapped_exec_fwd, wrapped_exec_bwd)
+    return wrapped_exec(params)
+
+
+# The execute function in forward mode
+def _execute_with_fwd(
+    params,
+    tapes=None,
+    device=None,
+    execute_fn=None,
+    gradient_kwargs=None,
+    _n=1,
+):  # pylint: disable=dangerous-default-value,unused-argument
+    @jax.custom_vjp
+    def wrapped_exec(params):
+        new_tapes = []
+
+        for t, a in zip(tapes, params):
+            new_tapes.append(t.copy(copy_operations=True))
+            new_tapes[-1].set_parameters(a)
+
+        with qml.tape.Unwrap(*new_tapes):
+            res, jacs = execute_fn(new_tapes, **gradient_kwargs)
+
+        if len(res) > 1:
+            res, jacs = [jnp.array(r) for r in res], [jnp.array(j) for j in jacs]
+        else:
+            res, jacs = jnp.array(res), jnp.array(jacs)
+        return res, jacs
+
+    def wrapped_exec_fwd(params):
+        res, jacs = wrapped_exec(params)
+        return res, tuple([jacs, params])
+
+    def wrapped_exec_bwd(params, g):
+
+        # Use the jacobian that was computed on the forward pass
+        jacs, params = params
+
+        # Note: there is an issue when jax.jacobian is being used, either due
+        # to issues with tensor updating (TypeError: Updates tensor must be of
+        # rank 0; got 1) or because jax.vmap introduces a redundant
+        # dimensionality in the result by duplicating entries
+        # Example to the latter:
+        #
+        # 1. Output when using jax.jacobian:
+        # DeviceArray([[-0.09983342,  0.01983384],\n
+        #              [-0.09983342, 0.01983384]], dtype=float64),
+        # DeviceArray([[ 0.        , -0.97517033],\n
+        #              [ 0.        , -0.97517033]], dtype=float64)),
+        #
+        # 2. Expected output:
+        # DeviceArray([[-0.09983342, 0.01983384],\n
+        #             [ 0.        , -0.97517033]]
+        #
+        # The output produced by this function matches 2.
+        scalar_outputs = all(t.output_dim == 1 for t in tapes)
+        if not scalar_outputs:
+            raise ValueError(
+                "Computing the jacobian of vector-valued tapes is not supported currently in forward mode."
+            )
```

Comment:
ah, this would mean no `qml.probs()` with `adjoint` :( Do you think there is a way of fixing this later?

Reply:
`qml.probs` is not supported by adjoint at the moment in general either :(

`QuantumFunctionError: Adjoint differentiation method does not support measurement probs`

I think this should be doable, it would just require more investigation once `qml.probs` support is provided for adjoint.